geoh5py/shared/merging/base.py

```diff
@@ -0,0 +1,200 @@
+        data_array[:] = np.nan
+
+        # populate the intermediate array
+        for data in all_data:
+            data_array[
+                unique_entity_types.index(data[1][:nb_keys]), data[-2] : data[-1]
+            ] = data[0]
+            unique_associations[data[1][:nb_keys]].append(data[2])
+
+        # create the output data
+        add_data_dict = {}
+        for id_, data in enumerate(data_array):
+            # define the association
+            association = unique_associations[unique_entity_types[id_]]
+
+            if not len(set(association)) == 1:
+                raise ValueError(
+                    f"Cannot merge data with different associations: {set(association)}"
+                )
+
+            # define the name
+            if nb_keys == 3:
+                name = f"merged_{unique_entity_types[id_][0].name}_{id_}"
+            elif nb_keys == 2:
+                name = f"merged_{unique_entity_types[id_][0].name}_{unique_entity_types[id_][1]}"
+            else:
+                name = f"merged_{unique_entity_types[id_][0].name}"
+
+            add_data_dict[name] = {
+                "association": association[0],
+                "values": data,
+                "entity_type": unique_entity_types[id_][0],
+            }
+
+        return add_data_dict
+
+    @classmethod
+    def merge_objects(
+        cls, input_entities: list[ObjectBase], add_data: bool = True, **kwargs
+    ):
+        """
+        Merge a list of :obj:geoh5py.objects.ObjectBase of the same type.
+        :param input_entities: the list of objects to merge.
+        :param add_data: if True, the data will be merged.
+        :param kwargs: the kwargs to create the new object.
+        :return: an object of type input_entities[0] containing the merged vertices.
+        """
+        cls.validate_objects(input_entities)
+
+        kwargs = cls.validate_kwargs(input_entities[0].workspace, **kwargs)
+
+        # create the vertices
+        vertices = np.vstack([input_entity.vertices for input_entity in input_entities])
+
+        # create an object of type input_entities[0]
+        output_type = type(input_entities[0])
+        output = output_type.create(vertices=vertices, **kwargs)
+
+        # add the data
+        if add_data:
+            output.add_data(cls.merge_data(input_entities))
+
+        return output
+
+    @classmethod
+    def validate_kwargs(cls, workspace: Workspace, **kwargs) -> dict:
+        """
+        Validate the kwargs to create a new object.
+        :param workspace: the workspace to eventually use.
+        :param kwargs: the kwargs to create the new object.
+        :return: the validated kwargs.
+        """
+
+        if "workspace" not in kwargs:
+            kwargs["workspace"] = workspace
+
+        return kwargs
+
+    @classmethod
+    @abstractmethod
+    def validate_type(cls, input_entity: ObjectBase):
+        raise NotImplementedError("BaseMerger cannot be used.")
```

Comment:
Don't need to raise anything since the @abstractmethod will generate an error on the child class if not implemented.
Can you add docstrings too.

Reply:
After a conversation here on this issue https://github.com/MiraGeoscience/geoh5py/pull/412#discussion_r1323606044,
it does make sense to raise NotImplementedError here, since it is technically reachable.